jdbc/src/main/scala/org/apache/spark/sql/sources/jdbcExtensions.scala

```diff
@@ -167,26 +192,86 @@ object JdbcExtendedUtils extends Logging {
   }
 
   def getTableWithSchema(table: String, conn: Connection,
-      session: Option[() => SnappySession] = None): (String, String) = {
+      session: Option[() => SparkSession] = None): (String, String) = {
     val dotIndex = table.indexOf('.')
     val schemaName = if (dotIndex > 0) {
       table.substring(0, dotIndex)
-    } else {
+    } else session match {
+      case None => conn.getSchema
       // get the current schema
-      if (session.isDefined) session.get().getCurrentSchema else conn.getSchema
+      case Some(s) => s().catalog.currentDatabase
     }
     val tableName = if (dotIndex > 0) table.substring(dotIndex + 1) else table
     (schemaName, tableName)
   }
 
-  private def getTableMetadataResultSet (table: String, conn: Connection): ResultSet = {
+  private def getTableMetadataResultSet(table: String, conn: Connection): ResultSet = {
     val (schemaName, tableName) = getTableWithSchema(table, conn)
-      // using the JDBC meta-data API
+    // using the JDBC meta-data API
     conn.getMetaData.getTables(null, schemaName, tableName, null)
   }
 
-  def tableExistsInMetaData(table: String, conn: Connection,
-      dialect: JdbcDialect): Boolean = {
+  protected lazy val getCatalystTypeMethod: Method = {
+    val m = JdbcUtils.getClass.getDeclaredMethod(
+      "org$apache$spark$sql$execution$datasources$jdbc$JdbcUtils$$getCatalystType",
+      classOf[Int], classOf[Int], classOf[Int], classOf[Boolean])
+    m.setAccessible(true)
+    m
+  }
+
+  /**
+   * Is the given JDBC type (java.sql.Types) data type signed.
+   */
+  def isSigned(typeId: Int): Boolean = {
+    typeId == Types.INTEGER || typeId == Types.FLOAT || typeId == Types.DECIMAL ||
+        typeId == Types.SMALLINT || typeId == Types.BIGINT || typeId == Types.TINYINT ||
+        typeId == Types.NUMERIC || typeId == Types.REAL || typeId == Types.DOUBLE
+  }
+
+  def getTableSchema(schemaName: String, tableName: String, conn: Connection): Seq[StructField] = {
+    val rs = conn.getMetaData.getColumns(null, schemaName, tableName, null)
+    if (rs.next()) {
+      val cols = new mutable.ArrayBuffer[StructField]()
+      do {
+        // COLUMN_NAME
+        val columnName = rs.getString(4)
+        // DATA_TYPE
+        val jdbcType = rs.getInt(5)
+        // TYPE_NAME
+        val typeName = rs.getString(6)
+        // COLUMN_SIZE
+        val size = rs.getInt(7)
```

Comment:
For String type what should be expected of size. Can we rely on this size method?

Reply:
It should be else its a bug.